File: app/src/main/java/com/kent/university/privelt/api/DataExtraction.java
package com.kent.university.privelt.api;

import android.content.Context;
import android.text.TextUtils;
import android.util.Log;

import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.api.client.extensions.android.http.AndroidHttp;
import com.google.api.client.googleapis.extensions.android.gms.auth.GoogleAccountCredential;
import com.google.api.client.json.gson.GsonFactory;
import com.google.api.services.drive.DriveScopes;
import com.kent.university.privelt.BuildConfig;
import com.kent.university.privelt.PriVELTApplication;
import com.kent.university.privelt.R;
import com.kent.university.privelt.database.PriVELTDatabase;
import com.kent.university.privelt.model.Service;
import com.kent.university.privelt.model.Settings;
import com.kent.university.privelt.model.UserData;
import com.kent.university.privelt.repositories.ServiceDataRepository;
import com.kent.university.privelt.repositories.SettingsDataRepository;
import com.kent.university.privelt.repositories.UserDataRepository;
import com.kent.university.privelt.utils.DriveServiceHelper;
import com.kent.university.webviewautologin.response.ResponseCallback;
import com.kent.university.webviewautologin.response.ResponseEnum;
import com.kent.university.webviewautologin.services.LoginService;
import com.university.kent.dataextractor.DataExtractor;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static com.kent.university.privelt.injections.Injection.provideServiceDataSource;
import static com.kent.university.privelt.injections.Injection.provideSettingsDataSource;
import static com.kent.university.privelt.injections.Injection.provideUserDataSource;
import static com.kent.university.privelt.model.UserData.DELIMITER;

public class DataExtraction {

    private static final String TAG = DataExtractor.class.getSimpleName();

    private static void saveToGoogleDrive() {

        ServiceDataRepository serviceDataRepository = provideServiceDataSource(PriVELTApplication.getInstance());

        SettingsDataRepository settingsDataRepository = provideSettingsDataSource(PriVELTApplication.getInstance());

        Settings settings = settingsDataRepository.getInstantSettings();

        if (settings != null && settings.isGoogleDriveAutoSave()) {
            try {
                GoogleSignInAccount account = GoogleSignIn.getLastSignedInAccount(PriVELTApplication.getInstance());

                GoogleAccountCredential credential =
                        GoogleAccountCredential.usingOAuth2(
                                PriVELTApplication.getInstance(), Collections.singleton(DriveScopes.DRIVE_FILE));
                credential.setSelectedAccount(account.getAccount());
                com.google.api.services.drive.Drive googleDriveService =
                        new com.google.api.services.drive.Drive.Builder(
                                AndroidHttp.newCompatibleTransport(),
                                new GsonFactory(),
                                credential)
                                .setApplicationName(PriVELTApplication.getInstance().getResources().getString(R.string.app_name))
                                .build();
                DriveServiceHelper mDriveServiceHelper = new DriveServiceHelper(googleDriveService);

                List<Service> serviceList = serviceDataRepository.getAllServices();
                List<Service> oldServices = cloneList(serviceList);
                for (int i = 0; i < serviceList.size(); i++) {
                    serviceList.get(i).setPassword("");
                    serviceList.get(i).setUser("");
                    serviceList.get(i).setPasswordSaved(false);
                    serviceDataRepository.updateServices(serviceList.get(i));
                }
                String s = mDriveServiceHelper.uploadFile(PriVELTApplication.getInstance().getDatabasePath(PriVELTDatabase.PriVELTDatabaseName), settings.getGoogleDriveFileID());
                for (Service service : oldServices)
                    serviceDataRepository.updateServices(service);

                settings.setGoogleDriveFileID(s);
                settingsDataRepository.updateSettings(settings);

                if (BuildConfig.DEBUG)
                    Log.d(TAG, "Google drive saved");

            }
            catch (Exception ignored) {

            }
        }
    }

    public static void processExtractionForEachService(Context applicationContext) {

        ServiceHelper serviceHelper = new ServiceHelper(((PriVELTApplication) (applicationContext)).getCurrentActivity());

        ServiceDataRepository serviceDataRepository = provideServiceDataSource(applicationContext);

        List<Service> services = serviceDataRepository.getAllServices();
        for (com.kent.university.privelt.model.Service service : services) {
            if (!service.isPasswordSaved())
                continue;
            processDataExtraction(serviceHelper, service, service.getUser(), service.getPassword(), applicationContext);
        }
     }

    private static List<Service> cloneList(List<Service> list) throws CloneNotSupportedException {
        List<Service> clone = new ArrayList<>(list.size());
        for (Service item : list) clone.add((Service) item.clone());
        return clone;
    }

    public static void processDataExtraction(ServiceHelper serviceHelper, com.kent.university.privelt.model.Service service, String email, String password, Context applicationContext) {

        UserDataRepository userDataRepository = provideUserDataSource(applicationContext);

        LoginService loginService = serviceHelper.getServiceWithName(service.getName());

        DataExtractor dataExtractor = new DataExtractor(loginService);
        final ArrayList<UserData> allUserData = new ArrayList<>();

        ((PriVELTApplication) (applicationContext)).getCurrentActivity().runOnUiThread(() -> loginService.autoLogin(email, password, new ResponseCallback() {
            @Override
            public void getResponse(ResponseEnum responseEnum, String data) {
                if (BuildConfig.DEBUG)
                    Log.d(TAG, responseEnum.toString());
                if (responseEnum == ResponseEnum.SUCCESS) {
                    dataExtractor.injectAll(((PriVELTApplication)applicationContext).getCurrentActivity(), (jsonArray, status) -> {
                        if (BuildConfig.DEBUG)
                            Log.d(TAG, status.toString());
                        if (jsonArray != null) {
                            allUserData.addAll(parseJSON(jsonArray, service));
                            if (BuildConfig.DEBUG)
                                Log.d(TAG, jsonArray.toString());
                        }
                        if (status.isDone()) {
                            if (BuildConfig.DEBUG)
                                Log.d(TAG, "LOGIN SERVICE:" + allUserData.size());
                            userDataRepository.deleteUserDatasForAService(service.getId());

                            for (UserData userData : allUserData)
                                userDataRepository.insertUserDatas(userData);

                            saveToGoogleDrive();
                        }
                    });
                }
            }
        }));
    }


    private static ArrayList<UserData> parseJSON(JSONArray jsonArray, com.kent.university.privelt.model.Service service) {
        ArrayList<UserData> array = new ArrayList<>();
        try {
            for (int i = 0; i < jsonArray.length(); i++) {
                JSONObject obj = jsonArray.getJSONObject(i);
                JSONArray datas = obj.getJSONArray("data");
                List<String> td = new ArrayList<>();
                for (int j = 0; j < datas.length(); j++) {
                    td.add(datas.getString(j));
                }
                UserData userData = new UserData(
                        obj.getString("title"),
                        obj.getString("type"),
                        obj.getString("value"),
                        TextUtils.join(DELIMITER, td),
                        service.getId());
                array.add(userData);
            }
        } catch (Exception ignored) {
        }
        return array;
    }
}

File: app/src/main/java/com/kent/university/privelt/api/DataExtraction.kt
/*
 *  This Source Code Form is subject to the terms of the Mozilla Public
 *  * License, v. 2.0. If a copy of the MPL was not distributed with this
 *  * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.kent.university.privelt.api

import android.content.Context
import android.text.TextUtils
import android.util.Log
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.api.client.extensions.android.http.AndroidHttp
import com.google.api.client.googleapis.extensions.android.gms.auth.GoogleAccountCredential
import com.google.api.client.json.gson.GsonFactory
import com.google.api.services.drive.Drive
import com.google.api.services.drive.DriveScopes
import com.kent.university.privelt.BuildConfig
import com.kent.university.privelt.PriVELTApplication
import com.kent.university.privelt.R
import com.kent.university.privelt.database.PriVELTDatabase
import com.kent.university.privelt.model.Service
import com.kent.university.privelt.model.UserData
import com.kent.university.privelt.utils.DriveServiceHelper
import net.neferett.webviewsextractor.DataExtractor
import net.neferett.webviewsextractor.extraction.Status
import net.neferett.webviewsinjector.response.ResponseCallback
import net.neferett.webviewsinjector.response.ResponseEnum
import org.json.JSONArray
import java.util.*

object DataExtraction {
    private val TAG = DataExtractor::class.simpleName
    private fun saveToGoogleDrive(applicationContext: Context) {
        val serviceDataRepository = PriVELTDatabase.getInstance(applicationContext)?.serviceDao()
        val settingsDataRepository = PriVELTDatabase.getInstance(applicationContext)?.settingsDao()
        val settings = settingsDataRepository?.instantSettings
        if (settings != null && settings.isGoogleDriveAutoSave) {
            try {
                val account = GoogleSignIn.getLastSignedInAccount(PriVELTApplication.instance)
                val credential = GoogleAccountCredential.usingOAuth2(
                        PriVELTApplication.instance, listOf(DriveScopes.DRIVE_FILE))
                credential.selectedAccount = account!!.account
                val googleDriveService = Drive.Builder(
                                AndroidHttp.newCompatibleTransport(),
                                GsonFactory(),
                                credential)
                        .setApplicationName(PriVELTApplication.instance?.resources?.getString(R.string.app_name))
                        .build()
                val mDriveServiceHelper = DriveServiceHelper(googleDriveService)
                val serviceList = serviceDataRepository?.allServices
                val oldServices = cloneList(serviceList!!)
                for (i in serviceList.indices) {
                    serviceList[i].password = ""
                    serviceList[i].user = ""
                    serviceList[i].isPasswordSaved = false
                    serviceDataRepository.updateServices(serviceList[i])
                }
                val s = mDriveServiceHelper.uploadFile(PriVELTApplication.instance?.getDatabasePath(PriVELTDatabase.PriVELTDatabaseName)!!, settings.googleDriveFileID)
                for (service in oldServices) serviceDataRepository.updateServices(service)
                settings.googleDriveFileID = s
                settingsDataRepository.updateSettings(settings)
                if (BuildConfig.DEBUG) Log.d(TAG, "Google drive saved")
            } catch (e: Exception) {
                if (BuildConfig.DEBUG) Log.d(TAG, "Google drive not saved " + e.message)
            }
        }
    }

    fun processExtractionForEachService(applicationContext: Context) {
        val serviceHelper = ServiceHelper((applicationContext as PriVELTApplication).currentActivity)
        val serviceDataRepository = PriVELTDatabase.getInstance(applicationContext)?.serviceDao()
        val services = serviceDataRepository?.allServices
        for (service in services!!) {
            if (service.isPasswordSaved)
                processDataExtraction(serviceHelper, service, service.user, service.password, applicationContext)
        }
    }

    private fun cloneList(list: List<Service>): List<Service> {
        val clone: MutableList<Service> = ArrayList(list.size)
        for (item in list) clone.add(item.clone() as Service)
        return clone
    }

    fun processDataExtraction(serviceHelper: ServiceHelper, service: Service, email: String?, password: String?, applicationContext: Context) {
        val time = System.currentTimeMillis()
        val userDataRepository = PriVELTDatabase.getInstance(applicationContext)?.userDataDao()
        val previousUserData = userDataRepository?.getUserDataForAService(service.id)
        val loginService = serviceHelper.getServiceWithName(service.name)
        val dataExtractor = DataExtractor(loginService)
        val allUserData = ArrayList<UserData>()
        (applicationContext as PriVELTApplication).currentActivity?.runOnUiThread {
            loginService?.autoLogin(email, password, object : ResponseCallback() {
                override fun getResponse(responseEnum: ResponseEnum, data: String) {
                    if (BuildConfig.DEBUG) Log.d(TAG + " FOR SERVICE=" + service.name, responseEnum.toString())
                    if (responseEnum == ResponseEnum.SUCCESS) {
                        dataExtractor.injectAll(applicationContext.currentActivity) { jsonArray: JSONArray?, status: Status ->
                            if (BuildConfig.DEBUG) Log.d(TAG + " FOR SERVICE=" + service.name, status.toString())
                            if (jsonArray != null) {
                                allUserData.addAll(parseJSON(jsonArray, service, time))
                                if (BuildConfig.DEBUG) Log.d(TAG + " FOR SERVICE=" + service.name, jsonArray.toString())
                            }
                            if (status.isDone) {
                                if (BuildConfig.DEBUG) Log.d(TAG + " FOR SERVICE=" + service.name, "LOGIN SERVICE:" + allUserData.size)
                                for (userData in allUserData) {
                                    //check if we have already extracted this data
                                    val res = previousUserData?.find {
                                        it.value == userData.value && it.title == userData.title
                                    }
                                    if (res == null)
                                        userDataRepository?.insertUserData(userData)
                                }
                                saveToGoogleDrive(applicationContext)
                            }
                        }
                    }
                }
            })
        }
    }

    private fun parseJSON(jsonArray: JSONArray, service: Service, time: Long): ArrayList<UserData> {
        val array = ArrayList<UserData>()
        try {
            for (i in 0 until jsonArray.length()) {
                val obj = jsonArray.getJSONObject(i)
                val data = obj.getJSONArray("data")
                val td: MutableList<String?> = ArrayList()
                for (j in 0 until data.length()) {
                    td.add(data.getString(j))
                }
                val userData = UserData(
                        obj.getString("title"),
                        obj.getString("type"),
                        obj.getString("value"),
                        TextUtils.join(UserData.DELIMITER, td),
                        service.id,
                        time)
                array.add(userData)
            }
        } catch (ignored: Exception) {
        }
        return array
    }
}